Fix planning failure due to incomplete pushdown

martint opened the pull request:
Given these tables:

```
CREATE TABLE t1(id integer);
CREATE TABLE t2(id bigint);
```
The following query fails with "Equi criteria are empty, so INNER join should not have PARTITIONED distribution type"

```
SELECT t1.id
FROM t1
JOIN t2 ON t1.id = t2.id
WHERE t1.id = 10
```

This occurs because PredicatePushdown cannot derive an equivalence during the first pass that helps
it infer that the join condition is always true. By the time the join distribution strategy is picked,
the join is still an equi-join, so the planner chooses PARTITIONED. Later on, during another pass of
PredicatePushdown, the condition is determined to be always true, and that implicitly converts the
Join to a cross join.

This change enhances PredicatePushdown to consider the predicate above the join to simplify the
effective predicate of the left/right branches of the join before engaging in the usual inference
logic. This is equivalent to inlining equalities defined by the predicate into the left/right
predicates and allows the equivalence inferencer to "infer more" and push predicates across the
branches of the join.

Fixes https://github.com/prestosql/presto/issues/2320

martint committed:
Fix planning failure due to incomplete pushdown

Given these tables:

  CREATE TABLE t1(id integer);
  CREATE TABLE t2(id bigint);

The following query fails with "Equi criteria are empty, so INNER join should not have PARTITIONED distribution type"

   SELECT t1.id
   FROM t1
   JOIN t2 ON t1.id = t2.id
   WHERE t1.id = 10

This occurs because PredicatePushdown cannot derive an equivalence during the first pass that helps
it infer that the join condition is always true. By the time the join distribution strategy is picked,
the join is still an equi-join, so the planner chooses PARTITIONED. Later on, during another pass of
PredicatePushdown, the condition is determined to be always true, and that implicitly converts the
Join to a cross join.

This change enhances PredicatePushdown to consider the predicate above the join to simplify the
effective predicate of the left/right branches of the join before engaging in the usual inference
logic. This is equivalent to inlining equalities defined by the predicate into the left/right
predicates and allows the equivalence inferencer to "infer more" and push predicates across the
branches of the join.

findepi commented on presto-testing/src/main/java/io/prestosql/testing/AbstractTestJoinQueries.java:
```diff
@@ -2256,6 +2256,25 @@ public void testRightJoinWithEmptyProbeSide()
+    @Test
+    public void testEquijoinOnDifferentTypesWithFilter()
+    {
+        // Non-sensical query that exercises a join on integer vs bigint with a filter on the integer column that can be pushed down
```
remove "Non-sensical". Most test queries do not make much sense

martint replied:
It’s particularly non-sensical because it joins nation with lineitem on keys that are unrelated. But fair enough, I’ll remove it.

sopel39 commented:
This seem to solve problem only partially.

> Later on, during another pass of PredicatePushdown, the condition is determined to be always true, and that implicitly converts the Join to a cross join.

I think this implicit conversion should not produce invalid join node, but rather change join type to replicated.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -885,14 +885,20 @@ private static InnerJoinPushDownResult processInnerJoin(
 
+            // Attempt to simplify the effective left/right predicates with the predicate we're pushing down
+            // This, effectively, inlines any constants derived from such predicate
+            EqualityInference predicateInference = EqualityInference.newInstance(metadata, inheritedPredicate);
```
Could we build more nested plan, where there are still not enough predicate pushdowns iterations?
See: https://github.com/prestodb/presto/issues/11569 (inference for outer joins might is limited and predicates might be derived only after n-th run depending how nested joins are)

kasiafi commented:
> I think this implicit conversion should not produce invalid join node, but rather change join type to replicated.

I think `PredicatePushDown` should be able to adjust the distribution type in the case when it produces a cross join.
However, this mustn't happen after `AddExchanges`, and `PredicatePushDown` is run once after `AddExchanges`.

sopel39 commented:
> However, this mustn't happen after AddExchanges, and PredicatePushDown is run once after AddExchanges.

Right, so maybe PP shouldn't touch predicate at all if they require join type change (it's not PP responsibility)

kasiafi commented:
> Right, so maybe PP shouldn't touch predicate at all if they require join type change (it's not PP responsibility)
I think we may want to keep this optimisation. Maybe PP could be aware if it is before or after AddExchanges?
Apart from that, we need some enhancement of PP to reduce the number of iterations. I checked the provided example query. It is a simple query, and still it needs exactly as many passes of PP to pushdown fully, as there are currently in PlanOptimizers.

sopel39 commented:
> I think we may want to keep this optimisation. Maybe PP could be aware if it is before or after AddExchanges?

If we do that before add exchanges, then it shouldn't matter as there is no PP execution between add exhanges and reorder joins/choose join distribution type

> Apart from that, we need some enhancement of PP to reduce the number of iterations.

this requires moving PP to iterative optimizer, but that requires more generic traits (for inherited predicate)

martint commented:
> I think this implicit conversion should not produce invalid join node, but rather change join type to replicated.

This points to a more fundamental problem with how the IR and optimization work today (it's I'm my long-term list of things to address as we continue to evolve the IR). A partitioned join *needs* an equijoin criteria. A cross join doesn't have one. Changing it to replicated would break other assumptions and decisions AddExchanges made, so it can't be done in isolation.

kasiafi commented:
> If we do that before add exchanges, then it shouldn't matter as there is no PP execution between add exhanges and reorder joins/choose join distribution type

So my suggestion is that we allow PredicatePushDown to produce cross joins, and set their DistributionType to REPLICATED or just leave it empty so that DetermineJoinDistributionType() will determine it.
The last point where we can do this is somewhere around the line 517.

There is one more occurrence of PredicatePushDown after AddExchanges. There we have to make sure that we don't produce a cross join from a PARTITIONED join.

martint committed:
Unwrap casts during predicate pushdown

The presence of projections of the form x = CAST(y AS t)
was causing pushdown to get blocked, which resulted in
needing more passes to fully pushdown certain predicates.

findepi commented on presto-main/src/test/java/io/prestosql/sql/planner/optimizations/TestReorderWindows.java:
```diff
@@ -248,7 +248,7 @@ public void testNotReorderAcrossNonPartitionFilter()
                                         .addFunction(functionCall("avg", commonFrame, ImmutableList.of(QUANTITY_ALIAS))),
                                 project(
-                                        filter(RECEIPTDATE_ALIAS + " IS NOT NULL",
+                                        filter("NOT (" + RECEIPTDATE_ALIAS + " IS NULL)",
```
btw why `IS NOT NULL` isn't the canonical form?

martint replied:
Because we want the IR to be as small as possible. We already need to support NOT as a fundamental boolean operation. Note that the plan testing framework “cheats” and uses SQL expressions as a way to describe IR, so the fact that it’s the canonical IR form doesn’t mean it’s the canonical form in the SQL language.

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/iterative/rule/UnwrapCastInComparison.java:
```diff
@@ -112,6 +112,15 @@ private static ExpressionRewriter createRewrite(Metadata metadata, TypeAnalyzer
         };
     }
 
+    public static Expression unwrapCasts(Session session, Metadata metadata, TypeAnalyzer typeAnalyzer, TypeProvider types, Expression expression)
```
use `unwrapCasts` in `createRewrite`

sopel39 commented on presto-main/src/main/java/io/prestosql/sql/planner/optimizations/PredicatePushDown.java:
```diff
@@ -258,6 +260,8 @@ public PlanNode visitProject(ProjectNode node, RewriteContext<Expression> contex
 
             List<Expression> inlinedDeterministicConjuncts = inlineConjuncts.get(true).stream()
                     .map(entry -> inlineSymbols(node.getAssignments().getMap(), entry))
+                    .map(conjunct -> canonicalizeExpression(conjunct, typeAnalyzer.getTypes(session, types, conjunct), metadata))
```
add comment why this is needed?
BTW: why canonicalization is needed here?

martint replied:
It's because UnwrapCasts only works with expressions of the form `CAST(x AS T) = constant`. `canonicalizeExpression` converts `constant = CAST(x AS T)`
